llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -592,27 +713,324 @@ static Error applyAddends(MCCASReader &Reader,
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      DWARFExtractor.getULEB128(&CUOffset, &Err);
+      FormSize += CUOffset - PrevOffset;
+      break;
+    }
+    case dwarf::DW_FORM_LLVM_addrx_offset: {
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      DWARFExtractor.getULEB128(&CUOffset, &Err);
+      FormSize += CUOffset - PrevOffset + 4;
+      break;
+    }
+    case dwarf::DW_FORM_string: {
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      auto CurrOffset = CUOffset;
+      DWARFExtractor.getCStr(&CUOffset, &Err);
+      FormSize += CUOffset - CurrOffset;
+      break;
+    }
+    case dwarf::DW_FORM_indirect: {
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      FormVal =
+          static_cast<dwarf::Form>(DWARFExtractor.getULEB128(&CUOffset, &Err));
+      Indirect = true;
+      FormSize += CUOffset - PrevOffset;
+      break;
+    }
+    case dwarf::DW_FORM_strp:
+    case dwarf::DW_FORM_sec_offset:
+    case dwarf::DW_FORM_GNU_ref_alt:
+    case dwarf::DW_FORM_GNU_strp_alt:
+    case dwarf::DW_FORM_line_strp:
+    case dwarf::DW_FORM_strp_sup: {
+      FormSize += FP.getDwarfOffsetByteSize();
+      break;
+    }
+    case dwarf::DW_FORM_addrx3:
+    case dwarf::DW_FORM_lo_user: {
+      llvm_unreachable("usupported form");
+      break;
+    }
+    }
+  } while (Indirect && !Err);
+
+  if (Err)
+    return std::move(Err);
+
+  return FormSize;
+}
+
+static Error materializeCUDie(DWARFCompileUnit &DCU,
+                              MutableArrayRef<char> SectionContents,
+                              StringRef CUData,
+                              ArrayRef<char> DistinctDataArrayRef,
+                              uint64_t &CUOffset, uint64_t &DistinctDataOffset,
+                              uint64_t &SectionOffset) {
+  // Copy Abbrev Tag.
+  DWARFDataExtractor DWARFExtractor(CUData, DCU.isLittleEndian(),
+                                    DCU.getAddressByteSize());
+  uint64_t PrevOffset = CUOffset;
+  uint64_t AbbrevTag = DWARFExtractor.getULEB128(&CUOffset);
+  std::memcpy(&SectionContents[SectionOffset], &AbbrevTag,
+              CUOffset - PrevOffset);
+  SectionOffset += CUOffset - PrevOffset;
+
+  // Return if the abbrev tag is 0, this indicates the end of a sequence of
+  // children DWARFDies
+  if (AbbrevTag == 0)
+    return Error::success();
+
+  // Create a empty DWARFDie to extract the attributes.
+  DWARFDebugInfoEntry DDIE;
+  auto *AbbrevDecl =
+      DCU.getAbbreviations()->getAbbreviationDeclaration(AbbrevTag);
+  DDIE.setAbbrevDecl(AbbrevDecl);
+  DWARFDie CUDie(&DCU, &DDIE);
+
+  // Loop over all attributes in a compile unit die, if it is a DW_FORM_strp or
+  // DW_FORM_sec_offset, read the data from the DistinctDataArrayRef, otherwise
```

Comment:
> if it is a DW_FORM_strp or DW_FORM_sec_offset

It is going to be very easy for this comment to be out of sync, maybe something like "reader the data from DistinctDataArrayRef or from the CU data, depending on the form of the attribute"